File: 3rd Sem/NM/C-program/gaussseidel.c
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

int main() {
    int n;
    printf("Enter matrix size: ");
    scanf("%d", &n);

    double **a = (double **)malloc(n * sizeof(double *));
    double *b = (double *)malloc(n * sizeof(double));
    double *x = (double *)malloc(n * sizeof(double));
    double *x_new = (double *)malloc(n * sizeof(double));
    double tol = 1e-6;

    for (int i = 0; i < n; i++) {
        a[i] = (double *)malloc(n * sizeof(double));
    }

    printf("Enter matrix row-wise:\n");
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            scanf("%lf", &a[i][j]);
        }
    }

    printf("Enter RHS values: ");
    for (int i = 0; i < n; i++) {
        scanf("%lf", &b[i]);
    }

    for (int i = 0; i < n; i++) {
        x[i] = 0; // Initial guess for solution
    }

    while (1) {
        for (int i = 0; i < n; i++) {
            double s = 0;
            for (int j = 0; j < n; j++) {
                if (j != i) {
                    s += a[i][j] * x[j];
                }
            }
            x_new[i] = (b[i] - s) / a[i][i];
        }

        int converged = 1;
        for (int i = 0; i < n; i++) {
            if (fabs(x_new[i] - x[i]) >= tol) {
                converged = 0;
                break;
            }
        }

        if (converged) {
            break;
        }

        for (int i = 0; i < n; i++) {
            x[i] = x_new[i];
        }
    }

    printf("Solution: ");
    for (int i = 0; i < n; i++) {
        printf("%.6f ", x[i]);
    }
    printf("\n");

    // Free dynamically allocated memory
    for (int i = 0; i < n; i++) {
        free(a[i]);
    }
    free(a);
    free(b);
    free(x);
    free(x_new);

    return 0;
}

File: 3rd Sem/NM/python/gaussseidel.py
n = int(input("Enter matrix size: "))  # Number of variables
a = [list(map(float, input(f"Enter row {i + 1}: ").split())) for i in range(n)]  # Coefficient matrix
b = list(map(float, input("Enter RHS values: ").split()))  # Right-hand side values
x = [0] * n  # Initial guess for the solution
tol = 1e-6  # Tolerance level for convergence

while True:
    x_new = x[:]  # Copy of the previous solution
    
    # Iterate through each row and update the solution
    for i in range(n):
        s = sum(a[i][j] * x_new[j] for j in range(n) if j != i)  # Sum of all terms except the diagonal
        x_new[i] = (b[i] - s) / a[i][i]  # Update the i-th value of the solution
    
    # Check for convergence: if the change is smaller than the tolerance, break the loop
    if all(abs(x_new[i] - x[i]) < tol for i in range(n)):
        break
    
    # Update x with the new values for the next iteration
    x = x_new

# Output the solution
print("Solution:", *x)
